bfvmm/include/serial/serial_port_base.h

```diff
@@ -0,0 +1,166 @@
+//
+// Bareflank Hypervisor
+// Copyright (C) 2017 Assured Information Security, Inc.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#ifndef SERIAL_PORT_BASE_H
+#define SERIAL_PORT_BASE_H
+
+#include <cstdint>
+
+#include <intrinsics/common.h>
+
+// -----------------------------------------------------------------------------
+// Exports
+// -----------------------------------------------------------------------------
+
+#include <bfexports.h>
+
+#ifndef STATIC_SERIAL
+#ifdef SHARED_SERIAL
+#define EXPORT_SERIAL EXPORT_SYM
+#else
+#define EXPORT_SERIAL IMPORT_SYM
+#endif
+#else
+#define EXPORT_SERIAL
+#endif
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4251)
+#endif
+
+// -----------------------------------------------------------------------------
+// Definitions
+// -----------------------------------------------------------------------------
+
+/// Serial port base class
+///
+/// This class provides an interface to the basic functions of a serial port,
+/// and also provides abstracted IO functions to its subclasses.
+///
+class EXPORT_SERIAL serial_port_base
+{
+public:
+#if defined(BF_X64)
+    using port_type = x64::portio::port_addr_type;
+    using value_type_8 = x64::portio::port_8bit_type;
+    using value_type_32 = x64::portio::port_32bit_type;
+#elif defined(BF_AARCH64)
+    using port_type = uintptr_t;
```

Comment:
uintptr_t changes in size, what that intended? Also, does it make sense to make aliases for these in the intrinsics so that other modules are using the same type?

Reply:
aarch64 uses MMIO for serial port IO, so the "port" type should be exactly the width of a pointer. As for making aliases in intrinsics - I don't think so, they're part of the abstracted IO specifically for use by the serial port drivers. Putting pieces of them in intrinsics seems like what I'm trying to remove in this commit.

Reply:
+1